### PR TITLE
Use gcs for queue-health storage, move to vm

### PR DESCRIPTION
**Pull request by fejta**
`start-vm.sh`: stages files, recreates the vm w/ a startup script and storage scope 
`startup.sh`: startup script that installs prerequisites and starts graph and poller programs
`poller.py`: poller program that downloads previous state, polls submit-queue to append new info and uploads state to gcs
`graph.py`: downloads the current state, renders it to a png file and saves to a public gcs object

Next step will be to create images for this, create a pod and move it into our utility cluster

cc @lavalamp

### CODE REVIEW

File: queue-health/graph.py [hunk 1]
```diff
@@ -18,25 +18,42 @@
 
 import collections
 import datetime
+import subprocess
+import sys
+import StringIO
```
Comment (rmmh) [inline queue-health/graph.py]: `import cStringIO as StringIO`

File: queue-health/graph.py [hunk 2]
```diff
@@ -18,25 +18,42 @@
...
-daily_happiness = []
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+import matplotlib.dates as mdates
 
-blocked_intervals = []
-daily_queue = collections.deque()
+def open_lines():
+  return open('results.txt', 'r')
```
Comment (rmmh) [inline queue-health/graph.py]: 4 space indents in this file

File: queue-health/graph.py [hunk 3]
```diff
@@ -18,25 +18,42 @@
...
-with open('results.txt', 'r') as f:
     last_blocked = None
     daily_sum = 0
-    for _ in xrange(60 * 24 * 21):
-        f.readline()
-    for line in f:
-        date, time, online, pr, queue, run, blocked = line.strip().split(' ')
+    for line in history_lines:
+        try:
+          date, time, online, pr, queue, run, blocked = line.strip().split(' ')
```
Comment (rmmh) [inline queue-health/graph.py]: indents

File: queue-health/poller.py [hunk 4]
```diff
@@ -14,43 +14,97 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from __future__ import print_function
-
 import datetime
 import requests
+import StringIO
```
Comment (rmmh) [inline queue-health/poller.py]: `import cStringIO as StringIO`

File: queue-health/poller.py [hunk 5]
```diff
@@ -14,43 +14,97 @@
...
-            print(data, file=f)
-            f.flush()
-            time.sleep(60)
+            except KeyError:
+                traceback.print_exc()
+            except IOError:
+                traceback.print_exc()
+                pass
+
+            print >>sys.stderr, 'Appending current status to history'
```
Comment (rmmh) [inline queue-health/poller.py]: I think it would be more interesting to do `print data` than this.

File: queue-health/startup.sh [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+set -o errexit
+
+HISTORY='gs://kubernetes-test-history/sq/history.txt'
+GRAPH='gs://kubernetes-test-history/k8s-queue-health.png'
+
+POLLER='gs://kubernetes-test-history/sq/poller.py'
+GRAPHER='gs://kubernetes-test-history/sq/graph.py'
+
+install-deps() {
+  for i in {1..5}; do
```
Comment (rmmh) [inline queue-health/startup.sh]: ???

Comment (fejta) [inline queue-health/startup.sh]: HAHAHA

File: queue-health/poller.py (hunk 4 above)
Comment (fejta) [inline queue-health/poller.py]: Done

File: queue-health/graph.py (hunk 3 above)
Comment (fejta) [inline queue-health/graph.py]: Done

File: queue-health/graph.py (hunk 2 above)
Comment (fejta) [inline queue-health/graph.py]: Done

File: queue-health/graph.py (hunk 1 above)
Comment (fejta) [inline queue-health/graph.py]: Done

File: queue-health/poller.py (hunk 5 above)
Comment (fejta) [inline queue-health/poller.py]: I don't think this tradeoff is worth it:

a) I don't want to print a 4MB file every minute
b) If I'm interested in the contents I can just gsutil cat the file (and don't process the file at all)

Comment (fejta) [inline queue-health/poller.py]: Nm, printing the new line is a great idea

File: jenkins/agent-ctl.sh [hunk 7]
```diff
@@ -169,9 +169,6 @@ auto-agent() {
   detatch-agent
```
Comment (rmmh) [inline jenkins/agent-ctl.sh]: typo: `detach`

File: jenkins/agent-ctl.sh [hunk 8]
```diff
@@ -39,12 +39,12 @@ usage() {
   echo '  update-image: update the image-family used to create new disks'
   echo 'Common commands:'
   echo '  # Refresh image'
-  echo '  $(basename "${0}") --base-image light-agent auto-image'
```
Comment (rmmh) [inline jenkins/agent-ctl.sh]: I don't think this is supposed to be in this commit?

**Comment by fejta** [comment]
@rmmh PTAL removed the unexpected file

File: queue-health/graph.py [hunk 9]
```diff
@@ -17,26 +17,43 @@
...
-queued = []
-instant_happiness = []
-daily_happiness = []
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+import matplotlib.dates as mdates
+
+def open_lines():
+  return open('results.txt', 'r')
```
Comment (rmmh) [inline queue-health/graph.py]: indent

File: queue-health/graph.py [hunk 10]
```diff
@@ -17,26 +17,43 @@
...
-prs = []
-queued = []
-instant_happiness = []
-daily_happiness = []
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+import matplotlib.dates as mdates
+
+def open_lines():
```
Comment (rmmh) [inline queue-health/graph.py]: this function is dead?

File: queue-health/graph.py [hunk 11]
```diff
@@ -69,47 +86,74 @@
...
+def render_forever(history_uri, png_uri):
+    """Download results from history_uri, render to png and save to png_uri."""
+    buf = cStringIO.StringIO()
+    while True:
+        print >>sys.stderr, 'Truncate render buffer'
+        buf.seek(0)
+        buf.truncate()
+        print >>sys.stderr, 'Cat latest results from %s...' % history_uri
+        try:
+          history = subprocess.check_output(['gsutil', 'cat', history_uri])
```
Comment (rmmh) [inline queue-health/graph.py]: 4-space indent

File: queue-health/poller.py [hunk 12]
```diff
@@ -14,43 +14,97 @@
...
+
 
 def get_submit_queue_json(path):
-    return requests.get('http://submit-queue.k8s.io/{}'.format(path)).json()
+    for n in range(3):
+        uri = 'http://submit-queue.k8s.io/%s' % path
+        print >>sys.stderr, 'GET %s' % uri
+        resp = requests.get(uri)
+        if resp.ok:
+          break
```
Comment (rmmh) [inline queue-health/poller.py]: indent

File: queue-health/poller.py [hunk 13]
```diff
@@ -14,43 +14,97 @@
...
+            buf.write(data)
+
+            print >>sys.stderr, 'Saving historical stats to %s...' % uri
+            save_stats(uri, buf)
         except KeyboardInterrupt:
             break
+
+
+if __name__ == '__main__':
+    poll_forever(*sys.argv[1:])
```
Comment (rmmh) [inline queue-health/poller.py]: using `f(*x[1:])` to cause an error if you have extra arguments is a little weird

File: queue-health/start-vm.sh [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+PREVIOUS="$(gcloud compute instances list --filter="name=${NAME}" || true)"
+if [[ -n "${PREVIOUS}" ]]; then
+  echo "Delete previous version of ${NAME}..."
+  gcloud -q compute instances delete "${NAME}"
+fi
+
+echo "Create ${NAME}..."
+gcloud compute instances create \
+  "${NAME}" \
+  --boot-disk-size=50GB \
```
Comment (rmmh) [inline queue-health/start-vm.sh]: could probably be 10GB

**Comment by rmmh** [comment]
LGTM

File: queue-health/graph.py (hunk 10 above)
Comment (fejta) [inline queue-health/graph.py]: Done

File: queue-health/graph.py (hunk 11 above)
Comment (fejta) [inline queue-health/graph.py]: Done

File: queue-health/poller.py (hunk 12 above)
Comment (fejta) [inline queue-health/poller.py]: Done

File: queue-health/poller.py (hunk 13 above)
Comment (fejta) [inline queue-health/poller.py]: Agreed it is not ideal but don't care given the small scope of the script.

File: queue-health/start-vm.sh (hunk 14 above)
Comment (fejta) [inline queue-health/start-vm.sh]: Hmm... I think I'd rather just move to GKE for the next version rather than try too hard to optimize the VM

**Pull request merged**